dubbo-config/dubbo-config-dynamic/src/main/java/org/apache/dubbo/config/dynamic/support/apollo/ApolloDynamicConfiguration.java

```diff
@@ -0,0 +1,134 @@
+    /**
+     * support two namespaces: application -> dubbo
+     */
+    private Config dubboConfig;
+    private Config appConfig;
+
+    public ApolloDynamicConfiguration() {
+
+    }
+
+    @Override
+    public void init() {
+        /**
+         * Instead of using Dubbo's configuration, I would suggest use the original configuration method Apollo provides.
+         */
+//        String configEnv = env.getCompositeConf().getString(ENV_KEY);
+//        String configCluster = env.getCompositeConf().getString(CLUSTER_KEY);
+        String configEnv = url.getParameter(Constants.CONFIG_ENV_KEY);
+        String configAddr = url.getParameter(Constants.CONFIG_ADDRESS_KEY);
+        String configCluster = url.getParameter(Constants.CONFIG_CLUSTER_KEY);
+        if (configEnv != null) {
+            System.setProperty(APOLLO_ENV_KEY, configEnv);
+        }
+        if (configAddr != null) {
+            System.setProperty(APOLLO_ADDR_KEY, configAddr);
+        }
+        if (configCluster != null) {
+            System.setProperty(APOLLO_CLUSTER_KEY, configCluster);
+        }
+
+        dubboConfig = ConfigService.getConfig("dubbo");
+        appConfig = ConfigService.getAppConfig();
+    }
+
+    @Override
+    public void addListener(String key, ConfigurationListener listener) {
+        this.appConfig.addChangeListener(new ApolloListener(listener));
+        this.dubboConfig.addChangeListener(new ApolloListener(listener));
+    }
+
+    @Override
+    public String getConfig(String key, String group, ConfigurationListener listener) {
+        this.appConfig.addChangeListener(new ApolloListener(listener));
+        this.dubboConfig.addChangeListener(new ApolloListener(listener));
+        return getInternalProperty(key, group, 0L);
+    }
+
+    @Override
+    protected String getInternalProperty(String key, String group, long timeout) {
+        String value = appConfig.getProperty(key, null);
+        if (value == null) {
+            value = dubboConfig.getProperty(key, null);
+        }
+
+        return value;
+    }
+
+    public ConfigChangeType getChangeType(PropertyChangeType changeType) {
+        if (changeType.equals(PropertyChangeType.DELETED)) {
+            return ConfigChangeType.DELETED;
+        }
+        return ConfigChangeType.MODIFIED;
+    }
+
+    private class ApolloListener implements ConfigChangeListener {
+        private ConfigurationListener listener;
+        private URL url;
+
+        public ApolloListener(ConfigurationListener listener) {
+            this(listener.getUrl(), listener);
+        }
+
+        public ApolloListener(URL url, ConfigurationListener listener) {
+            this.url = url;
+            this.listener = listener;
+        }
+
+        @Override
+        public void onChange(ConfigChangeEvent changeEvent) {
+            System.out.println("Changes for namespace " + changeEvent.getNamespace());
+            for (String key : changeEvent.changedKeys()) {
```

Comment:
next version of apollo-client will provide an api like `appConfig. addChangeListener(someListener, someInterestedKeys)` so that the listener will only be called when the interested keys are changed.

Right now, we could use ConfigChangeEvent#isChanged to check whether some key is changed in this event.

Reply:
I also think a more fine-grained subscription is necessary.
`ConfigChangeEvent#isChanged` seems more reasonable than the loop and check at present.